test/functional/ui/messages_spec.lua

```diff
@@ -3138,3 +3138,385 @@ it('pager works in headless mode with UI attached', function()
+            title = 'TestSuit',
+          },
+          history = true,
+          id = 1,
+          kind = 'progress',
+        },
+      },
+    })
+
+    assert_progress_autocmd({
+      text = { 'test-message-updated' },
+      percent = 50,
+      status = 'running',
+      title = 'TestSuit',
+      id = 1,
+      data = {},
+    }, 'Progress autocmd receives progress update')
+
+    -- progress event can filter by title
+    setup_autocmd('Special Title')
+    api.nvim_echo(
+      { { 'test-message-updated' } },
+      true,
+      { id = id, kind = 'progress', percent = 80, status = 'running' }
+    )
+    assert_progress_autocmd(nil, 'No progress message with Special Title yet')
+
+    api.nvim_echo(
+      { { 'test-message-updated' } },
+      true,
+      { id = id, kind = 'progress', title = 'Special Title', percent = 100, status = 'success' }
+    )
+    assert_progress_autocmd({
+      text = { 'test-message-updated' },
+      percent = 100,
+      status = 'success',
+      title = 'Special Title',
+      id = 1,
+      data = {},
+    }, 'Progress autocmd receives progress update')
+  end)
+
+  it('user-defined data in `data` field', function()
+    api.nvim_echo({ { 'test-message' } }, true, {
+      kind = 'progress',
+      title = 'TestSuit',
+      percent = 10,
+      status = 'running',
+      data = { test_attribute = 1 },
+    })
+
+    screen:expect({
+      grid = [[
+        ^                         |
+        {1:~                        }|*4
+      ]],
+      messages = {
+        {
+          content = { { 'test-message' } },
+          history = true,
+          id = 1,
+          kind = 'progress',
+          progress = {
+            percent = 10,
+            status = 'running',
+            title = 'TestSuit',
+          },
+        },
+      },
+    })
+    assert_progress_autocmd({
+      text = { 'test-message' },
+      percent = 10,
+      status = 'running',
+      title = 'TestSuit',
+      id = 1,
+      data = { test_attribute = 1 },
+    }, 'Progress autocmd receives progress messages')
+  end)
+
+  it('validates', function()
+    -- throws error if title, status, percent, data is used in non progress message
+    eq(
+      'title, status, percents and data fields can only be used with progress messages',
+      t.pcall_err(api.nvim_echo, { { 'test-message' } }, false, { title = 'TestSuit' })
+    )
+
+    eq(
+      'title, status, percents and data fields can only be used with progress messages',
+      t.pcall_err(api.nvim_echo, { { 'test-message' } }, false, { status = 'running' })
+    )
+
+    eq(
+      'title, status, percents and data fields can only be used with progress messages',
+      t.pcall_err(api.nvim_echo, { { 'test-message' } }, false, { percent = 10 })
+    )
+
+    eq(
+      'title, status, percents and data fields can only be used with progress messages',
+      t.pcall_err(api.nvim_echo, { { 'test-message' } }, false, { data = { tag = 'test' } })
```

Comment:
why are we restricting `data` to Progress messages? I guess because we don't have a `Message` event yet.

Reply:
Yeah,, for now, data doesn't go anywhere except for progress-event. if we add a general message event we can open it up for other kinds.